src/stan_math_backend/Stan_math_code_gen.ml

```diff
@@ -101,24 +102,38 @@ let maybe_templated_arg_types (args : Program.fun_arg_decl) =
       | true -> Some (sprintf "T%d__" i)
       | false -> None )
 
-let maybe_require_templates (names : string option list)
-    (args : Program.fun_arg_decl) =
-  let require_for_arg arg =
-    match trd3 arg with
-    | UnsizedType.URowVector -> "stan::require_row_vector_t"
-    | UVector -> "stan::require_col_vector_t"
-    | UMatrix -> "stan::require_eigen_matrix_dynamic_t"
-    (* NB: Not unwinding array types due to the way arrays of eigens are printed *)
-    | _ -> "stan::require_stan_scalar_t" in
-  List.map2_exn names args ~f:(fun name a ->
-      match name with
-      | Some t -> Some (Require (require_for_arg a, t))
-      | None -> None )
+let require_templates (names : string option list) (args : Program.fun_arg_decl)
+    =
+  let require_for_arg arg t =
+    let requires arg =
+      match arg with
+      | UnsizedType.URowVector ->
+          [ Require ("stan::require_row_vector_t", t)
+          ; Require ("stan::require_not_vt_complex", t) ]
```

Comment:
I think to put these all in one require we want to

1. Make all of these just logic checks like `is_row_vector` and replace `require_not_vt_complex` and `is_real_value_type`
2. concatenate them all into one `require_all_t<>`. So a check for some like the signature below would have the associated templates

```stan
matrix foo(complex_vector a, matrix b)
```

Would be something like 

```cpp
template <typename T1__, typename T2__, require_all_t<is_eigen_vector<T1__>, is_complex_vt<T1__>, is_eigen_matrix_dynamic<T2__>, is_real_vt<T2__>>* = nullptr>
```

Ftr I'm fine with it as is until we add that to the math library

Reply:
Do we have all the required template machinery to do this right now? I think that the `require_(not_)vt_complex` stuff is all generated by macros right?